auslib/db.py

```diff
@@ -1188,6 +1223,114 @@ def mergeUpdate(self, old_row, what, changed_by, transaction=None):
             self.log.debug("Merged %s into scheduled change '%s'", what, sc["sc_id"])
 
 
+class RequiredSignoffsTable(AUSTable):
+    """RequiredSignoffsTables store and validate information about what types
+    and how many signoffs are required for the data provided in
+    `decisionColumns`. Subclasses are required to create a Table with the
+    necessary columns, and add those columns names to `decisionColumns`.
+    When changes are made to a RequiredSignoffsTable, it will look at its own
+    rows to determine whether or not that change needs signoff."""
+
+    decisionColumns = []
+
+    def __init__(self, db, dialect):
+        self.table.append_column(Column("role", String(50), primary_key=True))
+        self.table.append_column(Column("signoffs_required", Integer, nullable=False))
+
+        super(RequiredSignoffsTable, self).__init__(db, dialect, scheduled_changes=True, scheduled_changes_kwargs={"conditions": ["time"]})
+
+    def getPotentialRequiredSignoffs(self, affected_rows, transaction=None):
+        potential_required_signoffs = []
+        for row in affected_rows:
+            if not row:
+                continue
+            where = {col: row[col] for col in self.decisionColumns}
+            potential_required_signoffs.extend(self.select(where=where, transaction=transaction))
+        return potential_required_signoffs
+
+    def validate(self, columns, transaction=None):
+        for col in self.decisionColumns:
+            if columns[col] is None:
+                raise ValueError("{} are required.".format(self.decisionColumns))
+
+        if transaction:
+            users_with_role, = transaction.execute(
+                self.db.permissions.user_roles.t.count().where(self.db.permissions.user_roles.role == columns["role"])
+            ).fetchone()
+        else:
+            users_with_role, = self.getEngine().execute(
+                self.db.permissions.user_roles.t.count().where(self.db.permissions.user_roles.role == columns["role"])
+            ).fetchone()
+        if users_with_role < columns["signoffs_required"]:
+            msg = ", ".join(self.decisionColumns)
+            raise ValueError("Cannot require {} signoffs for {} - only {} users hold the {} role".format(
+                columns["signoffs_required"], msg, users_with_role, columns["role"]
+            ))
+
+    def insert(self, changed_by, transaction=None, dryrun=False, signoffs=None, **columns):
+        self.validate(columns, transaction=transaction)
+
+        if not self.db.hasPermission(changed_by, "required_signoff", "create", transaction=transaction):
+            raise PermissionDeniedError("{} is not allowed to create new Required Signoffs.".format(changed_by))
+
+        if not dryrun:
+            potential_required_signoffs = self.getPotentialRequiredSignoffs([columns], transaction=transaction)
+            verify_signoffs(potential_required_signoffs, signoffs)
```

Comment:
Specifically, I'm concerned that we may start having more and more conditionals in Rules, Releases, etc. based on dryrun or not. Eventually this could be problematic because the more differences between dryrun mode and real run mode, the fewer issues it's able to catch. I think it's probably OK not to block on this for now, but something to watch out for in the future.

Separating out the permissions checks might help, you're right. We'd have to decide how much we care about the data validation that's done as a side effect, though. Most of this is already checked at the web layer (eg, validating a blob), but things like checking whether or not a Release is mapped to before deletion wouldn't be caught at scheduling time.